src/mono/mono/mini/simd-intrinsics.c

```diff
@@ -1264,6 +1266,27 @@ emit_sri_vector (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature *fsi
 		ins->inst_c1 = arg0_type;
 		return ins;
 	}
+	case SN_WidenLower:
+	case SN_WidenUpper: {
+#ifdef TARGET_ARM64
+		if (!is_element_type_primitive (fsig->params [0]))
+			return NULL;
+
+		int op = id == SN_WidenLower ? OP_XLOWER : OP_XUPPER;
```

Comment:
These are intrinsics only available for Vector128. You need to check if you are handling Vector128 or Vector64 here. Refer to CoreCLR implementation here https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L21282-L21320

Reply:
I started with the implementation that's in CoreCLR but that didn't work for some reason. On the other hand, this implementation works for both `Vector128` and `Vector64`:

`Vector64.Widen(Vector64.Create(1.1f, 2.2f))`:
```llvm
%xlower = shufflevector <2 x float> %simd_vtype, <2 x float> undef, <1 x i32> zeroinitializer
%arm64_fcvtl = fpext <1 x float> %xlower to <1 x double>
%xupper = shufflevector <2 x float> %simd_vtype, <2 x float> undef, <1 x i32> <i32 1>
%arm64_fcvtl1 = fpext <1 x float> %xupper to <1 x double>
```

`Vector64.Widen(Vector64.Create((sbyte)1))`:
```llvm
%xlower = shufflevector <8 x i8> %0, <8 x i8> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
%arm64_ushll = sext <4 x i8> %xlower to <4 x i16>
%xupper = shufflevector <8 x i8> %0, <8 x i8> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
%arm64_ushll1 = sext <4 x i8> %xupper to <4 x i16>
```

Do you think I should change the implementation anyway?

Reply:
I started by following the CoreCLR implementation and I had different implementations for `Vector128` and `Vector64` but it didn't work (assertion at mini-llvm.c:1391 failed) while this implementation works for both scenarios. I'll look into the code that LLVM generates and whether it doesn't fall back to inefficient instructions on Monday.

Reply:
From my testing the code in this PR works well and LLVM optimizes it correctly.

Here's the IR of `Vector64.Widen(Vector64<sbyte>)`:
```llvm
; Function Attrs: uwtable
define dso_local monocc i128 @"System.Runtime.Intrinsics.Vector64:Widen (System.Runtime.Intrinsics.Vector64`1<sbyte>)"([1 x i64] %arg_source) #0 gc "coreclr" {
BB0:
  %vreg_loc_42 = alloca i128, align 16
  %arg_source.fca.0.extract = extractvalue [1 x i64] %arg_source, 0
  %0 = bitcast i64 %arg_source.fca.0.extract to <8 x i8>
  %xlower = shufflevector <8 x i8> %0, <8 x i8> undef, <4 x i32> <i32 0, i32 1, i32 2, i32 3>
  %arm64_sshll = sext <4 x i8> %xlower to <4 x i16>
  %xupper = shufflevector <8 x i8> %0, <8 x i8> undef, <4 x i32> <i32 4, i32 5, i32 6, i32 7>
  %arm64_sshll1 = sext <4 x i8> %xupper to <4 x i16>
  %1 = bitcast i128* %vreg_loc_42 to i8*
  call void @llvm.memset.p0i8.i32(i8* nonnull align 16 dereferenceable(16) %1, i8 0, i32 16, i1 false)
  %2 = bitcast <4 x i16> %arm64_sshll to i64
  %3 = bitcast <4 x i16> %arm64_sshll1 to i64
  %4 = load void (i64*, [1 x i64], [1 x i64])*, void (i64*, [1 x i64], [1 x i64])** @"[tramp_1082] System.ValueTuple`2<System.Runtime.Intrinsics.Vector64`1<int16>, System.Runtime.Intrinsics.Vector64`1<int16>>:.ctor (System.Runtime.Intrinsics.Vector64`1<int16>,System.Runtime.Intrinsics.Vector64`1<int16>)", align 8
  %5 = bitcast i128* %vreg_loc_42 to i64*
  %.fca.0.insert = insertvalue [1 x i64] undef, i64 %2, 0
  %.fca.0.insert3 = insertvalue [1 x i64] undef, i64 %3, 0
  %6 = load i64, i64* inttoptr (i64 281473461021440 to i64*), align 4
  %7 = icmp eq i64 %6, 0
  br i1 %7, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i, !prof !1

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %8 = load void ()*, void ()** @mono_threads_state_poll, align 8
  call void %8() #2
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  notail call monocc void %4(i64* nonnull %5, [1 x i64] %.fca.0.insert, [1 x i64] %.fca.0.insert3), !managed_name !2
  %vreg_loc_48.sroa.0.0.copyload = load i128, i128* %vreg_loc_42, align 16
  ret i128 %vreg_loc_48.sroa.0.0.copyload
}
```

and here's IR for `Vector128.Widen(Vector128<sbyte>)`:
```llvm
; Function Attrs: uwtable
define dso_local monocc void @"System.Runtime.Intrinsics.Vector128:Widen (System.Runtime.Intrinsics.Vector128`1<sbyte>)"(%"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>"* noalias sret %vret, [2 x i64] %arg_source) #0 gc "coreclr" {
BB0:
  %vreg_loc_43 = alloca %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>", align 8
  %vreg_loc_34 = alloca i64, align 8
  %arg_source.fca.0.extract = extractvalue [2 x i64] %arg_source, 0
  %0 = bitcast i64 %arg_source.fca.0.extract to <8 x i8>
  %arg_source.fca.1.extract = extractvalue [2 x i64] %arg_source, 1
  %1 = bitcast i64 %arg_source.fca.1.extract to <8 x i8>
  %2 = ptrtoint %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>"* %vret to i64
  store i64 %2, i64* %vreg_loc_34, align 8
  %3 = alloca %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>", align 8
  %arm64_sshll = sext <8 x i8> %0 to <8 x i16>
  %arm64_sshll1 = sext <8 x i8> %1 to <8 x i16>
  %4 = bitcast %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>"* %vreg_loc_43 to i8*
  call void @llvm.memset.p0i8.i32(i8* nonnull align 8 dereferenceable(32) %4, i8 0, i32 32, i1 false)
  %5 = load void (i64*, [2 x i64], [2 x i64])*, void (i64*, [2 x i64], [2 x i64])** @"[tramp_1144] System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>:.ctor (System.Runtime.Intrinsics.Vector128`1<int16>,System.Runtime.Intrinsics.Vector128`1<int16>)", align 8
  %6 = getelementptr inbounds %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>", %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>"* %vreg_loc_43, i64 0, i32 0
  %llvm_outarg_vt.0.vec.extract.bc = bitcast <8 x i16> %arm64_sshll to <2 x i64>
  %llvm_outarg_vt.0.vec.extract.extract = extractelement <2 x i64> %llvm_outarg_vt.0.vec.extract.bc, i32 0
  %.fca.0.insert = insertvalue [2 x i64] undef, i64 %llvm_outarg_vt.0.vec.extract.extract, 0
  %llvm_outarg_vt.8.vec.extract.extract = extractelement <2 x i64> %llvm_outarg_vt.0.vec.extract.bc, i32 1
  %.fca.1.insert = insertvalue [2 x i64] %.fca.0.insert, i64 %llvm_outarg_vt.8.vec.extract.extract, 1
  %llvm_outarg_vt2.0.vec.extract.bc = bitcast <8 x i16> %arm64_sshll1 to <2 x i64>
  %llvm_outarg_vt2.0.vec.extract.extract = extractelement <2 x i64> %llvm_outarg_vt2.0.vec.extract.bc, i32 0
  %.fca.0.insert3 = insertvalue [2 x i64] undef, i64 %llvm_outarg_vt2.0.vec.extract.extract, 0
  %llvm_outarg_vt2.8.vec.extract.extract = extractelement <2 x i64> %llvm_outarg_vt2.0.vec.extract.bc, i32 1
  %.fca.1.insert4 = insertvalue [2 x i64] %.fca.0.insert3, i64 %llvm_outarg_vt2.8.vec.extract.extract, 1
  %7 = load i64, i64* inttoptr (i64 281473535367936 to i64*), align 4
  %8 = icmp eq i64 %7, 0
  br i1 %8, label %gc.safepoint_poll.exit, label %gc.safepoint_poll.poll.i, !prof !1

gc.safepoint_poll.poll.i:                         ; preds = %BB0
  %9 = load void ()*, void ()** @mono_threads_state_poll, align 8
  call void %9() #2
  br label %gc.safepoint_poll.exit

gc.safepoint_poll.exit:                           ; preds = %BB0, %gc.safepoint_poll.poll.i
  notail call monocc void %5(i64* nonnull %6, [2 x i64] %.fca.1.insert, [2 x i64] %.fca.1.insert4), !managed_name !2
  %10 = bitcast %"System.ValueTuple`2<System.Runtime.Intrinsics.Vector128`1<int16>, System.Runtime.Intrinsics.Vector128`1<int16>>"* %3 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i32(i8* nonnull align 8 dereferenceable(32) %10, i8* nonnull align 8 dereferenceable(32) %4, i32 32, i1 false)
  %vreg_loc_34.0.vreg_loc_34.0.vreg_loc_34.0. = load volatile i64, i64* %vreg_loc_34, align 8
  %11 = inttoptr i64 %vreg_loc_34.0.vreg_loc_34.0.vreg_loc_34.0. to i8*
  call void @llvm.memcpy.p0i8.p0i8.i32(i8* nonnull align 1 dereferenceable(32) %11, i8* nonnull align 8 dereferenceable(32) %10, i32 32, i1 false)
  ret void
}
```

I'm not sure if the resulting LLVM IR can be improved any further. From what I can tell most of the code assembles the `ValueTuple`.